cpp/include/raft/neighbors/cagra_types.hpp

```diff
@@ -0,0 +1,197 @@
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the indices in the source dataset
+ *
+ */
+template <typename T, typename IdxT>
+struct index : ann::index {
+  static_assert(!raft::is_narrowing_v<uint32_t, IdxT>,
+                "IdxT must be able to represent all values of uint32_t");
+
+ public:
+  /** Distance metric used for clustering. */
+  [[nodiscard]] constexpr inline auto metric() const noexcept -> raft::distance::DistanceType
+  {
+    return metric_;
+  }
+
+  // /** Total length of the index. */
+  [[nodiscard]] constexpr inline auto size() const noexcept -> IdxT { return dataset_.extent(0); }
+
+  /** Dimensionality of the data. */
+  [[nodiscard]] constexpr inline auto dim() const noexcept -> uint32_t
+  {
+    return dataset_.extent(1);
+  }
+  /** Graph degree */
+  [[nodiscard]] constexpr inline auto graph_degree() const noexcept -> uint32_t
+  {
+    return graph_.extent(1);
+  }
+
+  /** Dataset [size, dim] */
+  [[nodiscard]] inline auto dataset() const noexcept -> device_matrix_view<const T, IdxT, row_major>
+  {
+    return dataset_.view();
+  }
+
+  /** neighborhood graph [size, graph-degree] */
+  inline auto graph() noexcept -> device_matrix_view<IdxT, IdxT, row_major>
+  {
+    return graph_.view();
+  }
+
+  [[nodiscard]] inline auto graph() const noexcept
+    -> device_matrix_view<const IdxT, IdxT, row_major>
+  {
+    return graph_.view();
+  }
+
+  // Don't allow copying the index for performance reasons (try avoiding copying data)
+  index(const index&) = delete;
+  index(index&&)      = default;
+  auto operator=(const index&) -> index& = delete;
+  auto operator=(index&&) -> index& = default;
+  ~index()                          = default;
+
+  /** Construct an empty index. */
+  index(raft::device_resources const& res)
+    : ann::index(),
+      metric_(raft::distance::DistanceType::L2Expanded),
+      dataset_(make_device_matrix<T, IdxT>(res, 0, 0)),
+      graph_(make_device_matrix<IdxT, IdxT>(res, 0, 0))
+  {
+  }
+
+  /** Construct an index from dataset and knn_graph arrays */
+  template <typename data_accessor, typename graph_accessor>
+  index(raft::device_resources const& res,
+        raft::distance::DistanceType metric,
+        mdspan<const T, matrix_extent<IdxT>, row_major, data_accessor> dataset,
+        mdspan<IdxT, matrix_extent<IdxT>, row_major, graph_accessor> knn_graph)
+    : ann::index(),
+      metric_(metric),
+      dataset_(make_device_matrix<T, IdxT>(res, dataset.extent(0), dataset.extent(1))),
+      graph_(make_device_matrix<IdxT, IdxT>(res, knn_graph.extent(0), knn_graph.extent(1)))
+  {
+    RAFT_EXPECTS(dataset.extent(0) == knn_graph.extent(0),
+                 "Dataset and knn_graph must have equal number of rows");
+    raft::copy(dataset_.data_handle(), dataset.data_handle(), dataset.size(), res.get_stream());
+    raft::copy(graph_.data_handle(), knn_graph.data_handle(), knn_graph.size(), res.get_stream());
```

Comment:
Are these copies necessary if the dataset is already on device?

Reply:
Yes, I think we can go about this a couple of ways:
1. Always copy, as is currently happening
2. Copy just the view, which would be like maintaining a reference to the underlying data
3. Construct the index on the fly if it's sufficiently cheap
4. Let the user pick which option to use from above. For example, if the user can guarantee that the data will stay in scope then I think they might go with 2

Could you please add this to your follow-on tasks as well? We can chat more later about how we want to go about this.

Reply:
Thanks for the suggestions! I have added an item to to #1392